File: packages/ui/Sources/UI/Toast/ToastHost.swift
import AlertToast
import SwiftUI

/// Mounts the app-wide AlertToast presenter over `content`, driven by a `ToastController`. AlertToast's
/// HUD uses system material, so it follows the device light/dark theme automatically.
public struct ToastHost<Content: View>: View {
  @ObservedObject private var controller: ToastController

  private let content: Content

  public init(controller: ToastController, @ViewBuilder content: () -> Content) {
    self.controller = controller
    self.content = content()
  }

  public var body: some View {
    content
      .toast(isPresenting: $controller.isPresenting, duration: 2.5) {
        AlertToast(
          displayMode: .hud,
          type: alertType(for: controller.tone),
          title: controller.message)
      }
  }

  private func alertType(for tone: ToastTone) -> AlertToast.AlertType {
    switch tone {
    case .success: .complete(.green)
    case .warning: .systemImage("exclamationmark.triangle.fill", .orange)
    case .danger: .error(.red)
    case .neutral: .regular
    }
  }
}

File: packages/ui/Java/com/variance/nearby/ui/ToastHost.kt
package com.variance.nearby.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Modifier
import com.dokar.sonner.ToastType
import com.dokar.sonner.Toaster
import com.dokar.sonner.rememberToasterState

/**
 * Mounts the app-wide compose-sonner toaster over [content], driven by [controller]. The toaster
 * follows the device light/dark theme.
 */
@Composable
fun ToastHost(
    controller: ToastController,
    content: @Composable () -> Unit,
) {
    val toaster = rememberToasterState()
    LaunchedEffect(controller) {
        controller.events.collect { event ->
            toaster.show(event.message, type = event.tone.toToastType())
        }
    }
    Box(modifier = Modifier.fillMaxSize()) {
        content()
        Toaster(state = toaster, darkTheme = isSystemInDarkTheme())
    }
}

private fun ToastTone.toToastType(): ToastType =
    when (this) {
        ToastTone.SUCCESS -> ToastType.Success
        ToastTone.WARNING -> ToastType.Warning
        ToastTone.DANGER -> ToastType.Error
        ToastTone.NEUTRAL -> ToastType.Normal
    }
